presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -1424,6 +1435,56 @@ else if (materializedViewStatus.isPartiallyMaterialized()) {
             return SqlFormatterUtil.getFormattedSql(unionQuery, sqlParser, Optional.empty());
         }
 
+        /**
+         * Returns materialized view status, considering filter conditions from query
+         */
+        private MaterializedViewStatus getMaterializedViewStatus(QualifiedObjectName materializedViewName, Optional<Scope> scope)
+        {
+            TupleDomain<String> baseQueryDomain;
+            // If there is a filter in the QuerySpecification we are currently visiting which has been rewritten by a materialized view,
+            // We can avoid counting partitions which will not be present in the final query when considering if we fail the optimization
+            // Due to too many missing partitions.
+            // TODO: Collect predicates in outer queries and apply the union to this criteria. i.e. in
+            //  SELECT x FROM (SELECT x FROM tbl WHERE y < 20) WHERE y > 5 where y is a partition key,
+            //  right now we would only remove partitions y >= 20. We should eventually consider y <= 5 as well.
+            if (currentSubquery.getWhere().isPresent()) {
```

Comment:
Can we create the config for this feature (where we consider base query domain) we should be able to disable this if required.